File: engine/financeCalculator.ts
import { Decimal } from 'decimal.js';

const TWO_PLACES = 2;

function roundCurrency(val: Decimal | null | undefined): Decimal | null {
  if (!val) return null;
  return val.toDecimalPlaces(TWO_PLACES, Decimal.ROUND_HALF_UP);
}

export function formatINR(n: number): string {
  const rounded = Math.round(n);
  const s = rounded.toString();
  if (s.length <= 3) return '₹' + s;
  const last3 = s.slice(-3);
  const rest = s.slice(0, -3);
  const grouped = rest.replace(/\B(?=(\d{2})+(?!\d))/g, ',');
  return '₹' + grouped + ',' + last3;
}

export function calculateFinancialAssessment(inputs: any, policy: any): any {
  if (!policy || !policy.minimum_required_dscr) throw new Error("CRITICAL: minimum_required_dscr missing from policy");
  if (!policy.maximum_household_debt_ratio) throw new Error("CRITICAL: maximum_household_debt_ratio missing from policy");
  
  const result: any = {};
  
  const reqLoan = inputs.requested_loan_amount != null ? new Decimal(inputs.requested_loan_amount) : null;
  
  const rate = inputs.annual_interest_rate_percent != null ? new Decimal(inputs.annual_interest_rate_percent) : null;
  const repayMonths = inputs.repayment_tenure_months != null ? Number(inputs.repayment_tenure_months) : null;
  const moratorium = inputs.moratorium_months != null ? Number(inputs.moratorium_months) : 0;
  const moratoriumMethod = inputs.moratorium_interest_method || "NONE";
  
  const unitsSold = inputs.monthly_units_sold != null ? new Decimal(inputs.monthly_units_sold) : null;
  const price = inputs.selling_price_per_unit != null ? new Decimal(inputs.selling_price_per_unit) : null;
  const varCost = inputs.variable_cost_per_unit != null ? new Decimal(inputs.variable_cost_per_unit) : null;
  
  const missingFields: string[] = [];
  const checkMissing = (field: string) => {
    if (inputs[field] == null && inputs.monthly_fixed_cost == null) {
      missingFields.push(field);
      return new Decimal(0);
    }
    return new Decimal(inputs[field] || 0);
  };

  const labour = checkMissing('monthly_labour_cost');
  const rent = checkMissing('monthly_rent');
  const transport = checkMissing('monthly_transport_cost');
  const otherFixed = checkMissing('monthly_other_fixed_cost');
  
  if (missingFields.length > 0 && inputs.monthly_fixed_cost == null) {
    result.loan_structure_status = "INSUFFICIENT_DATA";
    result.business_affordability_status = "INSUFFICIENT_DATA";
    result.overall_readiness = "INCOMPLETE";
    result.missing_fields = missingFields;
    return result;
  }

  const fixedCost = inputs.monthly_fixed_cost != null ? new Decimal(inputs.monthly_fixed_cost) : labour.add(rent).add(transport).add(otherFixed);
  result.monthly_fixed_cost = Number(roundCurrency(fixedCost)!.toFixed(2));
  
  if (price != null && varCost != null) {
    const cm = price.sub(varCost);
    result.unit_contribution_margin = Number(roundCurrency(cm)!.toFixed(2));
    
    if (cm.greaterThan(0)) {
      result.break_even_units = Math.round(Number(fixedCost.div(cm))); 
      result.break_even_status = "VIABLE";
    } else if (cm.equals(0)) {
      result.break_even_units = null;
      result.break_even_status = "NO_FINITE_BREAK_EVEN";
    } else {
      result.break_even_units = null;
      result.break_even_status = "STRUCTURALLY_UNVIABLE";
    }
  }

  if (unitsSold != null && price != null && varCost != null) {
    const rev = price.mul(unitsSold);
    const vc = varCost.mul(unitsSold);
    const surplus = rev.sub(vc).sub(fixedCost);
    
    result.monthly_revenue = Number(roundCurrency(rev)!.toFixed(2));
    result.monthly_variable_cost = Number(roundCurrency(vc)!.toFixed(2));
    result.monthly_operating_surplus = Number(roundCurrency(surplus)!.toFixed(2));
    result.business_cash_available_for_debt_service = Number(roundCurrency(surplus)!.toFixed(2));
  }
  
  function getPCap(p: Decimal, r: Decimal, m: number, method: string): Decimal {
    if (method === "NONE" || m === 0) return p;
    if (method === "SIMPLE_CAPITALIZE") return p.mul(new Decimal(1).add(r.mul(m)));
    if (method === "COMPOUND_CAPITALIZE") return p.mul(new Decimal(1).add(r).pow(m));
    return p;
  }
  
  function getEmi(pCap: Decimal, r: Decimal, n: number): Decimal {
    if (n === 0) return new Decimal(0);
    if (r.equals(0)) return pCap.div(n);
    const onePlusRPow = new Decimal(1).add(r).pow(n);
    return pCap.mul(r).mul(onePlusRPow).div(onePlusRPow.sub(1));
  }
  
  function getPFromEmi(emi: Decimal, r: Decimal, n: number): Decimal {
    if (n === 0) return new Decimal(0);
    if (r.equals(0)) return emi.mul(n);
    const onePlusRPow = new Decimal(1).add(r).pow(n);
    return emi.mul(onePlusRPow.sub(1)).div(r.mul(onePlusRPow));
  }
  
  function getPFromPCap(pCap: Decimal, r: Decimal, m: number, method: string): Decimal {
    if (method === "NONE" || m === 0) return pCap;
    if (method === "SIMPLE_CAPITALIZE") return pCap.div(new Decimal(1).add(r.mul(m)));
    if (method === "COMPOUND_CAPITALIZE") return pCap.div(new Decimal(1).add(r).pow(m));
    return pCap;
  }
  
  const monthlyRate = rate != null ? rate.div(100).div(12) : null;
  const maxScheme = inputs.maximum_scheme_loan_amount != null ? new Decimal(inputs.maximum_scheme_loan_amount) : null;
  
  let candLoan = reqLoan;
  if (reqLoan != null && maxScheme != null) {
      candLoan = Decimal.min(reqLoan, maxScheme);
  }
  
  let emiVal = null;
  if (candLoan != null && monthlyRate != null && repayMonths != null) {
    const pCap = getPCap(candLoan, monthlyRate, moratorium, moratoriumMethod);
    emiVal = getEmi(pCap, monthlyRate, repayMonths);
    
    result.capitalized_principal = Number(roundCurrency(pCap)!.toFixed(2));
    result.candidate_emi = Number(roundCurrency(emiVal)!.toFixed(2));
    result.emi = Number(roundCurrency(emiVal)!.toFixed(2));
    
    const totalRepayment = emiVal.mul(repayMonths);
    const totalInt = totalRepayment.sub(candLoan);
    if (totalInt.greaterThanOrEqualTo(0)) {
      result.total_interest = Number(roundCurrency(totalInt)!.toFixed(2));
    }
  }

  let bizDscr = null;
  if (result.business_cash_available_for_debt_service != null && emiVal != null && emiVal.greaterThan(0)) {
    bizDscr = new Decimal(result.business_cash_available_for_debt_service).div(emiVal);
    result.business_dscr = Number(roundCurrency(bizDscr)!.toFixed(2));
  }

  
  const minDscr = new Decimal(policy.minimum_required_dscr);
  const maxHhDebt = new Decimal(policy.maximum_household_debt_ratio);

  let maxEmi = null;
  if (result.business_cash_available_for_debt_service != null) {
    maxEmi = new Decimal(result.business_cash_available_for_debt_service).div(minDscr);
    result.maximum_affordable_emi = Number(roundCurrency(maxEmi)!.toFixed(2));
  }

  let affP = null;
  if (maxEmi != null && maxEmi.greaterThan(0) && monthlyRate != null && repayMonths != null) {
    const affPCap = getPFromEmi(maxEmi, monthlyRate, repayMonths);
    affP = getPFromPCap(affPCap, monthlyRate, moratorium, moratoriumMethod);
    result.affordable_loan_amount = Number(roundCurrency(affP)!.toFixed(2));
  }
  
  if (reqLoan != null && affP != null) {
    let recLoan = Decimal.min(reqLoan, affP);
    if (maxScheme != null) {
      recLoan = Decimal.min(recLoan, maxScheme);
    }
    result.recommended_loan_amount = Number(roundCurrency(recLoan)!.toFixed(2));
  }
  
  if (reqLoan != null && (repayMonths == null || rate == null)) {
    result.loan_structure_status = "INSUFFICIENT_DATA";
    result.business_affordability_status = "INSUFFICIENT_DATA";
    const missing = [];
    if (repayMonths == null) missing.push("repayment_tenure_months");
    if (rate == null) missing.push("annual_interest_rate_percent");
    result.missing_fields = missing;
  }
  
  let hhPostRatio = null;
  if (!("monthly_household_nonbusiness_income" in inputs) || inputs.monthly_household_nonbusiness_income === null ||
      !("monthly_household_essential_expenses" in inputs) || inputs.monthly_household_essential_expenses === null ||
      !("existing_monthly_household_debt_payments" in inputs) || inputs.existing_monthly_household_debt_payments === null) {
    result.household_affordability_status = "INSUFFICIENT_DATA";
    result.overall_readiness = "INCOMPLETE";
    if (!result.missing_fields) result.missing_fields = [];
    if (!("monthly_household_nonbusiness_income" in inputs) || inputs.monthly_household_nonbusiness_income === null) result.missing_fields.push("monthly_household_nonbusiness_income");
    if (!("monthly_household_essential_expenses" in inputs) || inputs.monthly_household_essential_expenses === null) result.missing_fields.push("monthly_household_essential_expenses");
    if (!("existing_monthly_household_debt_payments" in inputs) || inputs.existing_monthly_household_debt_payments === null) result.missing_fields.push("existing_monthly_household_debt_payments");
  } else {
    const hhIncomeRaw = inputs.monthly_household_nonbusiness_income;
    const hhInc = new Decimal(hhIncomeRaw);
    const hhExp = new Decimal(inputs.monthly_household_essential_expenses);
    const hhDebt = new Decimal(inputs.existing_monthly_household_debt_payments);
    
    const hhFree = hhInc.sub(hhExp).sub(hhDebt);
    result.household_free_cash = Number(roundCurrency(hhFree)!.toFixed(2));
    
    if (hhInc.equals(0)) {
      result.household_existing_debt_ratio = null;
      result.household_buffer_ratio = null;
      result.household_income_status = "ZERO_INCOME";
      result.household_affordability_status = "HIGH_RISK";
    } else {
      const exRatio = hhDebt.div(hhInc);
      const bufRatio = hhFree.div(hhInc);
      result.household_existing_debt_ratio = Number(roundCurrency(exRatio)!.toFixed(2));
      result.household_buffer_ratio = Number(roundCurrency(bufRatio)!.toFixed(2));
      result.household_income_status = "VALID";
    }
    
    if (emiVal != null) {
      const postDebt = hhDebt.add(emiVal);
      result.post_loan_monthly_debt_payments = Number(roundCurrency(postDebt)!.toFixed(2));
      if (hhInc.equals(0)) {
        result.post_loan_household_debt_ratio = null;
      } else {
        hhPostRatio = postDebt.div(hhInc);
        result.post_loan_household_debt_ratio = Number(roundCurrency(hhPostRatio)!.toFixed(2));
      }
    }
  }
  
  if (bizDscr != null && hhPostRatio != null) {
    if (bizDscr.greaterThanOrEqualTo(minDscr)) {
      result.business_affordability_status = "READY_FOR_FINANCE_REVIEW";
    } else {
      result.business_affordability_status = "HIGH_RISK";
    }
    
    if (hhPostRatio.lessThanOrEqualTo(maxHhDebt)) {
      result.household_affordability_status = "READY_FOR_FINANCE_REVIEW";
    } else {
      result.household_affordability_status = "HIGH_RISK";
    }
    
    if (result.business_affordability_status === "READY_FOR_FINANCE_REVIEW" && result.household_affordability_status === "READY_FOR_FINANCE_REVIEW") {
      result.overall_readiness = "READY_FOR_FINANCE_REVIEW";
    } else {
      result.overall_readiness = "HIGH_RISK";
    }
  }
  
  return result;
}

File: backend/app/services/finance_calculator.py
"""
Canonical deterministic financial assessment engine.

This module exposes:
- Stage-level pure helpers for isolated mathematical testing
- calculate_financial_assessment() for full production derivation chain

Production callers MUST NOT supply pre-computed stage values.
Stage helpers are exposed for golden-test parity only.
"""
from decimal import Decimal, ROUND_HALF_UP, getcontext
from typing import Dict, Any, Optional, List
import hashlib
import json

getcontext().prec = 28
TWO_PLACES = Decimal("0.01")


def round_currency(value: Decimal) -> Decimal:
    if value is None:
        return None
    return value.quantize(TWO_PLACES, rounding=ROUND_HALF_UP)


# ── Stage-level pure helpers (exposed for golden-test parity) ──


def calculate_capitalized_principal(
    principal: Decimal, monthly_rate: Decimal, moratorium_months: int, method: str
) -> Decimal:
    """Capitalize principal through moratorium period."""
    if method == "NONE" or moratorium_months == 0:
        return principal
    elif method == "SIMPLE_CAPITALIZE":
        return principal * (Decimal("1") + monthly_rate * Decimal(moratorium_months))
    elif method == "COMPOUND_CAPITALIZE":
        return principal * ((Decimal("1") + monthly_rate) ** moratorium_months)
    return principal


def calculate_emi(capitalized_principal: Decimal, monthly_rate: Decimal, repayment_months: int) -> Decimal:
    """Standard EMI formula."""
    if repayment_months == 0:
        return Decimal("0")
    if monthly_rate == Decimal("0"):
        return capitalized_principal / Decimal(repayment_months)
    r = monthly_rate
    n = repayment_months
    return capitalized_principal * r * ((Decimal("1") + r) ** n) / (((Decimal("1") + r) ** n) - Decimal("1"))


def calculate_affordable_principal_from_emi(
    max_emi: Decimal, monthly_rate: Decimal, repayment_months: int,
    moratorium_months: int = 0, moratorium_method: str = "NONE"
) -> Decimal:
    """Inverse EMI: given max affordable EMI, compute max affordable principal."""
    if repayment_months == 0:
        return Decimal("0")
    if monthly_rate == Decimal("0"):
        p_cap = max_emi * Decimal(repayment_months)
    else:
        r = monthly_rate
        n = repayment_months
        p_cap = max_emi * (((Decimal("1") + r) ** n) - Decimal("1")) / (r * ((Decimal("1") + r) ** n))

    # Reverse moratorium capitalization to get original principal
    if moratorium_method == "NONE" or moratorium_months == 0:
        return p_cap
    elif moratorium_method == "SIMPLE_CAPITALIZE":
        return p_cap / (Decimal("1") + monthly_rate * Decimal(moratorium_months))
    elif moratorium_method == "COMPOUND_CAPITALIZE":
        return p_cap / ((Decimal("1") + monthly_rate) ** moratorium_months)
    return p_cap


def calculate_household_post_loan_ratio(
    household_income: Decimal, existing_debt: Decimal, emi: Decimal
) -> Optional[Decimal]:
    """Post-loan household debt ratio."""
    if household_income == Decimal("0"):
        return None
    return (existing_debt + emi) / household_income


def evaluate_readiness_from_ratios(
    business_dscr: Optional[Decimal],
    post_loan_household_debt_ratio: Optional[Decimal],
    min_dscr: Decimal,
    max_hh_debt: Decimal,
) -> Dict[str, str]:
    """Evaluate readiness status from pre-computed ratios."""
    result = {}
    if business_dscr is not None:
        if business_dscr >= min_dscr:
            result["business_affordability_status"] = "READY_FOR_FINANCE_REVIEW"
        else:
            result["business_affordability_status"] = "HIGH_RISK"

    if post_loan_household_debt_ratio is not None:
        if post_loan_household_debt_ratio <= max_hh_debt:
            result["household_affordability_status"] = "READY_FOR_FINANCE_REVIEW"
        else:
            result["household_affordability_status"] = "HIGH_RISK"

    biz = result.get("business_affordability_status")
    hh = result.get("household_affordability_status")
    if biz and hh:
        if biz == "READY_FOR_FINANCE_REVIEW" and hh == "READY_FOR_FINANCE_REVIEW":
            result["overall_readiness"] = "READY_FOR_FINANCE_REVIEW"
        else:
            result["overall_readiness"] = "HIGH_RISK"

    return result


def compute_input_hash(inputs: Dict[str, Any], policy_version: str) -> str:
    """SHA-256 hash of canonical normalized inputs + policy version."""
    canonical = {}
    for k in sorted(inputs.keys()):
        v = inputs[k]
        if v is not None:
            canonical[k] = str(v)
    canonical["_policy_version"] = policy_version
    return hashlib.sha256(json.dumps(canonical, sort_keys=True).encode()).hexdigest()


# ── Full production calculator ──


def calculate_financial_assessment(
    inputs: Dict[str, Any],
    policy: Dict[str, Any],
    allow_stage_overrides: bool = False,
) -> Dict[str, Any]:
    """
    Full deterministic financial assessment.

    When allow_stage_overrides=False (production default):
      All values are derived from root inputs. Pre-computed stage values are ignored.
    When allow_stage_overrides=True (golden test harness only):
      Pre-computed values like maximum_affordable_emi, candidate_emi,
      business_dscr, post_loan_household_debt_ratio are accepted as inputs.
    """
    result: Dict[str, Any] = {}
    missing_fields: List[str] = []

    # ── Extract root business inputs ──
    req_loan = _to_decimal_or_none(inputs.get("requested_loan_amount"))
    rate = _to_decimal_or_none(inputs.get("annual_interest_rate_percent"))
    repay_months = inputs.get("repayment_tenure_months")
    moratorium = inputs.get("moratorium_months", 0) or 0
    moratorium_method = inputs.get("moratorium_interest_method", "NONE")

    units_sold = inputs.get("monthly_units_sold")
    price = inputs.get("selling_price_per_unit")
    var_cost = inputs.get("variable_cost_per_unit")

    # ── Fixed costs: missing is NOT zero ──
    if "monthly_fixed_cost" in inputs and inputs["monthly_fixed_cost"] is not None:
        fixed_cost = Decimal(str(inputs["monthly_fixed_cost"]))
    else:
        # Sum only EXPLICITLY provided components; missing → track
        fixed_components = {
            "monthly_labour_cost": inputs.get("monthly_labour_cost"),
            "monthly_rent": inputs.get("monthly_rent"),
            "monthly_transport_cost": inputs.get("monthly_transport_cost"),
            "monthly_other_fixed_cost": inputs.get("monthly_other_fixed_cost"),
        }
        provided = {k: v for k, v in fixed_components.items() if v is not None}
        if len(provided) == 0 and units_sold is not None:
            # If business inputs exist but NO fixed cost components, it's missing data
            fixed_cost = None
        else:
            fixed_cost = sum(Decimal(str(v)) for v in provided.values()) if provided else Decimal("0")

    if fixed_cost is not None:
        result["monthly_fixed_cost"] = round_currency(fixed_cost)

    # ── Break-even analysis ──
    if price is not None and var_cost is not None:
        cm = Decimal(str(price)) - Decimal(str(var_cost))
        result["unit_contribution_margin"] = round_currency(cm)
        if cm > Decimal("0") and fixed_cost is not None:
            be = fixed_cost / cm
            result["break_even_units"] = int(be.to_integral_value(rounding=ROUND_HALF_UP))
            result["break_even_status"] = "VIABLE"
        elif cm == Decimal("0"):
            result["break_even_units"] = None
            result["break_even_status"] = "NO_FINITE_BREAK_EVEN"
        else:
            result["break_even_units"] = None
            result["break_even_status"] = "STRUCTURALLY_UNVIABLE"

    # ── Scenario handling ──
    scenario = inputs.get("scenario")
    if scenario and units_sold is not None and price is not None and var_cost is not None and fixed_cost is not None:
        sim_units = Decimal(str(units_sold))
        sim_price = Decimal(str(price))
        sim_var = Decimal(str(var_cost))
        sim_fixed = fixed_cost
        
        if scenario == "DEMAND_DROP_20":
            sim_units = sim_units * Decimal("0.8")
            result["simulated_monthly_units_sold"] = int(sim_units.to_integral_value(rounding=ROUND_HALF_UP))
        elif scenario == "RAW_MATERIAL_UP_20":
            sim_var = sim_var * Decimal("1.2")
            result["simulated_monthly_units_sold"] = int(sim_units.to_integral_value(rounding=ROUND_HALF_UP))
            result["simulated_variable_cost_per_unit"] = round_currency(sim_var)
        elif scenario == "TRANSPORT_COST_SPIKE_50":
            sim_trans = Decimal(str(inputs.get("monthly_transport_cost", 0) or 0)) * Decimal("1.5")
            old_trans = Decimal(str(inputs.get("monthly_transport_cost", 0) or 0))
            sim_fixed = sim_fixed - old_trans + sim_trans
            result["simulated_monthly_units_sold"] = int(sim_units.to_integral_value(rounding=ROUND_HALF_UP))
            result["simulated_monthly_transport_cost"] = round_currency(sim_trans)

        sim_rev = sim_units * sim_price
        sim_vc = sim_units * sim_var
        sim_surplus = sim_rev - (sim_vc + sim_fixed)
        
        result["simulated_monthly_revenue"] = round_currency(sim_rev)
        result["simulated_monthly_variable_cost"] = round_currency(sim_vc)
        result["simulated_monthly_fixed_cost"] = round_currency(sim_fixed)
        result["simulated_monthly_operating_surplus"] = round_currency(sim_surplus)

    # ── Business CFADS ──
    cfads = None
    if units_sold is not None and price is not None and var_cost is not None:
        if fixed_cost is None:
            missing_fields.append("monthly_fixed_cost")
        else:
            rev = Decimal(str(units_sold)) * Decimal(str(price))
            vc = Decimal(str(units_sold)) * Decimal(str(var_cost))
            surplus = rev - (vc + fixed_cost)

            result["monthly_revenue"] = round_currency(rev)
            result["monthly_variable_cost"] = round_currency(vc)
            result["monthly_operating_surplus"] = round_currency(surplus)
            result["business_cash_available_for_debt_service"] = round_currency(surplus)
            cfads = surplus

    # ── Monthly rate ──
    monthly_rate = None
    if rate is not None:
        monthly_rate = (rate / Decimal("100")) / Decimal("12")

    # ── Scheme cap ──
    max_scheme = _to_decimal_or_none(inputs.get("maximum_scheme_loan_amount"))
    if max_scheme is not None:
        result["maximum_scheme_loan_amount"] = round_currency(max_scheme)

    # ── Candidate loan ──
    if req_loan is not None:
        cand_loan = min(req_loan, max_scheme) if max_scheme is not None else req_loan
    else:
        cand_loan = None

    # ── EMI calculation ──
    emi = None
    if cand_loan is not None and monthly_rate is not None and repay_months is not None:
        p_cap = calculate_capitalized_principal(cand_loan, monthly_rate, moratorium, moratorium_method)
        emi = calculate_emi(p_cap, monthly_rate, repay_months)

        result["capitalized_principal"] = round_currency(p_cap)
        result["candidate_emi"] = round_currency(emi)
        result["emi"] = round_currency(emi)

        total_repayment = emi * Decimal(repay_months)
        total_int = total_repayment - cand_loan
        if total_int >= Decimal("0"):
            result["total_interest"] = round_currency(total_int)

    # Stage override: candidate_emi (golden test only)
    if allow_stage_overrides and "candidate_emi" in inputs and inputs["candidate_emi"] is not None:
        emi = Decimal(str(inputs["candidate_emi"]))

    # ── DSCR ──
    biz_dscr = None
    if cfads is not None and emi is not None and emi > 0:
        biz_dscr = cfads / emi
        result["business_dscr"] = round_currency(biz_dscr)

    # Stage override: business_dscr (golden test only)
    if allow_stage_overrides and "business_dscr" in inputs and inputs["business_dscr"] is not None:
        biz_dscr = Decimal(str(inputs["business_dscr"]))
        result["business_dscr"] = round_currency(biz_dscr)

    # ── Policy thresholds ──
    min_dscr = Decimal(str(policy.get("minimum_required_dscr", "1.25")))
    max_hh_debt = Decimal(str(policy.get("maximum_household_debt_ratio", "0.50")))

    # ── Max affordable EMI ──
    max_emi = None
    if cfads is not None:
        if cfads <= Decimal("0"):
            max_emi = Decimal("0")
            result["maximum_affordable_emi"] = Decimal("0.00")
            result["business_affordability_status"] = "HIGH_RISK"
        else:
            max_emi = cfads / min_dscr
            result["maximum_affordable_emi"] = round_currency(max_emi)

    # Stage override: maximum_affordable_emi (golden test only)
    if allow_stage_overrides and "maximum_affordable_emi" in inputs and inputs["maximum_affordable_emi"] is not None:
        max_emi = Decimal(str(inputs["maximum_affordable_emi"]))
        result["maximum_affordable_emi"] = round_currency(max_emi)

    # ── Affordable loan amount ──
    aff_p = None
    if max_emi is not None and max_emi > 0 and monthly_rate is not None and repay_months is not None:
        aff_p = calculate_affordable_principal_from_emi(
            max_emi, monthly_rate, repay_months, moratorium, moratorium_method
        )
        result["affordable_loan_amount"] = round_currency(aff_p)
    elif max_emi is not None and max_emi <= Decimal("0") and monthly_rate is not None:
        result["affordable_loan_amount"] = Decimal("0.00")

    # Stage override: affordable_loan_amount (golden test only)
    if allow_stage_overrides and "affordable_loan_amount" in inputs and inputs["affordable_loan_amount"] is not None:
        aff_p = Decimal(str(inputs["affordable_loan_amount"]))
        result["affordable_loan_amount"] = round_currency(aff_p)

    # ── Recommended loan ──
    if req_loan is not None and aff_p is not None:
        rec_loan = min(req_loan, aff_p)
        if max_scheme is not None:
            rec_loan = min(rec_loan, max_scheme)
        result["recommended_loan_amount"] = round_currency(rec_loan)

    # ── Readiness states (missing loan terms) ──
    if req_loan is not None and (repay_months is None or rate is None):
        result["loan_structure_status"] = "INSUFFICIENT_DATA"
        result["business_affordability_status"] = "INSUFFICIENT_DATA"
        if repay_months is None:
            missing_fields.append("repayment_tenure_months")
        if rate is None:
            missing_fields.append("annual_interest_rate_percent")

    # ── Household ──
    hh_income_raw = inputs.get("monthly_household_nonbusiness_income")
    hh_post_ratio = None

    if hh_income_raw is None:
        result["household_affordability_status"] = "INSUFFICIENT_DATA"
        if "overall_readiness" not in result:
            result["overall_readiness"] = "INCOMPLETE"
        # Golden test parity: do not append to missing_fields if we are just testing loan math
        if not allow_stage_overrides or "monthly_household_nonbusiness_income" in inputs:
            missing_fields.append("monthly_household_nonbusiness_income")
    else:
        hh_inc = Decimal(str(hh_income_raw))

        hh_exp_raw = inputs.get("monthly_household_essential_expenses")
        hh_debt_raw = inputs.get("existing_monthly_household_debt_payments")

        if hh_exp_raw is None:
            missing_fields.append("monthly_household_essential_expenses")
        if hh_debt_raw is None:
            missing_fields.append("existing_monthly_household_debt_payments")

        hh_exp = Decimal(str(hh_exp_raw)) if hh_exp_raw is not None else Decimal("0")
        hh_debt = Decimal(str(hh_debt_raw)) if hh_debt_raw is not None else Decimal("0")

        hh_free = hh_inc - hh_exp - hh_debt
        result["household_free_cash"] = round_currency(hh_free)

        if hh_inc == Decimal("0"):
            result["household_existing_debt_ratio"] = None
            result["household_buffer_ratio"] = None
            result["household_income_status"] = "ZERO_INCOME"
            result["household_affordability_status"] = "HIGH_RISK"
        else:
            ex_ratio = hh_debt / hh_inc
            buf_ratio = hh_free / hh_inc
            result["household_existing_debt_ratio"] = round_currency(ex_ratio)
            result["household_buffer_ratio"] = round_currency(buf_ratio)
            result["household_income_status"] = "VALID"

        if emi is not None:
            post_debt = hh_debt + emi
            result["post_loan_monthly_debt_payments"] = round_currency(post_debt)
            if hh_inc == Decimal("0"):
                result["post_loan_household_debt_ratio"] = None
            else:
                hh_post_ratio = post_debt / hh_inc
                result["post_loan_household_debt_ratio"] = round_currency(hh_post_ratio)

    # Stage override: post_loan_household_debt_ratio (golden test only)
    if allow_stage_overrides and "post_loan_household_debt_ratio" in inputs and inputs["post_loan_household_debt_ratio"] is not None:
        hh_post_ratio = Decimal(str(inputs["post_loan_household_debt_ratio"]))
        result["post_loan_household_debt_ratio"] = round_currency(hh_post_ratio)

    # ── Final readiness evaluation ──
    if biz_dscr is not None and hh_post_ratio is not None:
        readiness = evaluate_readiness_from_ratios(biz_dscr, hh_post_ratio, min_dscr, max_hh_debt)
        result.update(readiness)

    if missing_fields:
        result["missing_fields"] = missing_fields

    return result


def _to_decimal_or_none(val) -> Optional[Decimal]:
    if val is None:
        return None
    return Decimal(str(val))
